stdNullPtr/Dynamic-Email-Generator
# Implement Service

## Overview
Initial service layer implementation, alongside a language interpreter that can interpret our custom expression language.
Integration tests were also implemented.
Refer to #9 for examples.

## Related Issues
Resolves #9 - Implement Controller

## Testing Instructions
1. Checkout the branch and build the project with `mvn clean package`.
2. Follow the instructions in the README, to setup certificates ("Certificates" section in README)
3. Follow the instructions in the README, to run the service locally ("Running the docker-compose bundle locally" section in README)
4. Use the following CURL command to test the endpoint (refer to #2 for query param syntax):
   ```bash
   curl https://localhost:9443/app/v1/email/generate?str1=Ivan&str2=Millenium&str3=.com&expression=lit(str1)%20;%20raw(@)%20;%20last(str2,%203)%20;%20substr(str3,1,3)%20;%20lit(str3)&str1=Falcon&str1=Petar&str1=Ivo&str1=Luke&str3=domain
5. You should get the following response: 
    ```json
    {
        "data": [
            "Ivan@iumco.com",
            "Ivan@iumomdomain",
            "Falcon@iumco.com",
            "Falcon@iumomdomain",
            "Petar@iumco.com",
            "Petar@iumomdomain",
            "Ivo@iumco.com",
            "Ivo@iumomdomain",
            "Luke@iumco.com",
            "Luke@iumomdomain"
        ],
        "message": null
    }

The curl command translates to these inputs: 
str1=Ivan
str2=Millenium
str3=.com
expression=lit(str1) ; raw(@) ; last(str2, 3) ; substr(str3,1,3) ; lit(str3)
str1=Falcon
str1=Petar
str1=Ivo
str1=Luke
str3=domain
